### PR TITLE
global settings role to burn for all

### PR DESCRIPTION
**Pull request by sasurobert**
global settings role to burn for all

### CODE REVIEW

**Review by dorin-iancu — APPROVED** [review]

File: vm/systemSmartContracts/esdt.go [hunk 1]
```diff
@@ -1073,6 +1083,83 @@ func (e *esdt) togglePause(args *vmcommon.ContractCallInput, builtInFunc string)
...
+
+func (e *esdt) setBurnForAll(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	token, returnCode := e.checkInputReturnDataBurnForAll(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	burnForAllExists := checkIfDefinedRoleExistsInToken(token, []byte(vmcommon.ESDTRoleBurnForAll))
+	if burnForAllExists {
+		e.eei.AddReturnMessage("cannot set burn  for all as it was already set")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1119]: done

File: vm/systemSmartContracts/esdt_test.go [hunk 2]
```diff
@@ -4625,4 +4631,146 @@ func registerAndSetAllRolesWithTypeCheck(t *testing.T, typeArgument []byte, expe
...
+	burnForAllRole := &ESDTRoles{Roles: [][]byte{[]byte(vmcommon.ESDTRoleBurnForAll)}, Address: []byte{}}
+	token.SpecialRoles = append(token.SpecialRoles, burnForAllRole)
+
+	marshalizedData, _ := args.Marshalizer.Marshal(token)
+	tokensMap[string(tokenName)] = marshalizedData
+	eei.storageUpdate[string(eei.scAddress)] = tokensMap
+
+	vmInput.CallerAddr = owner
+	vmInput.Arguments = [][]byte{tokenName}
+	output = e.Execute(vmInput)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt_test.go:4750]: linter: `ineffectual assignment to output (ineffassign)`

File: vm/systemSmartContracts/esdt.go [hunk 3]
```diff
@@ -2050,6 +2160,9 @@ func (e *esdt) EpochConfirmed(epoch uint32, _ uint64) {
 
 	e.flagRegisterAndSetAllRoles.SetValue(epoch >= e.registerAndSetAllRolesEnableEpoch)
 	log.Debug("ESDT register and set all roles", "enabled", e.flagRegisterAndSetAllRoles.IsSet())
+
+	e.flagBurnForAll.SetValue(epoch >= e.burnForAllEnableEpoch)
+	log.Debug("ESDT register and set all roles", "enabled", e.flagBurnForAll.IsSet())
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:2165]: wrong log
`log.Debug("ESDT burn for all", "enabled", e.flagBurnForAll.IsSet())` *

File: vm/systemSmartContracts/esdt.go [hunk 4]
```diff
@@ -1073,6 +1083,83 @@ func (e *esdt) togglePause(args *vmcommon.ContractCallInput, builtInFunc string)
...
+		return vmcommon.UserError
+	}
+
+	esdtTransferData := builtInCall + "@" + hex.EncodeToString(tokenID)
+	e.eei.SendGlobalSettingToAll(e.eSDTSCAddress, []byte(esdtTransferData))
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) setBurnForAll(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:1111]: can you rename it to `setBurnRoleForAll` ?

File: vm/systemSmartContracts/esdt.go [hunk 5]
```diff
@@ -1073,6 +1083,83 @@ func (e *esdt) togglePause(args *vmcommon.ContractCallInput, builtInFunc string)
...
+	token, returnCode := e.checkInputReturnDataBurnForAll(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	burnForAllExists := checkIfDefinedRoleExistsInToken(token, []byte(vmcommon.ESDTRoleBurnForAll))
+	if burnForAllExists {
+		e.eei.AddReturnMessage("cannot set burn for all as it was already set")
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:1119]: `cannot set burn for all role as it was already set`

File: vm/systemSmartContracts/esdt.go [hunk 6]
```diff
@@ -1073,6 +1083,83 @@ func (e *esdt) togglePause(args *vmcommon.ContractCallInput, builtInFunc string)
...
+
+func (e *esdt) unSetBurnForAll(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	token, returnCode := e.checkInputReturnDataBurnForAll(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	burnForAllExists := checkIfDefinedRoleExistsInToken(token, []byte(vmcommon.ESDTRoleBurnForAll))
+	if !burnForAllExists {
+		e.eei.AddReturnMessage("cannot un set burn for all as it was not set")
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:1141]: `cannot unset burn for all role as it was not set`

File: vm/systemSmartContracts/esdt.go (hunk 4 above)
Comment (bogdan-rosianu) [reply vm/systemSmartContracts/esdt.go:1111]: or `setBurnRoleGlobally`

**Review by gabi-vuls — DISMISSED** [review]
System test passed. 
@@                            Log scanner                                  @@
##                            burn-for-all                                 ##
================================================================================
- Known Warnings              14
- New Warnings                4
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================

File: vm/systemSmartContracts/esdt.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:2165]: done

File: vm/systemSmartContracts/esdt.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1111]: done

File: vm/systemSmartContracts/esdt.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1119]: done

File: vm/systemSmartContracts/esdt.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1141]: done

File: vm/systemSmartContracts/esdt_test.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt_test.go:4750]: done

**Review by gabi-vuls — DISMISSED** [review]
System test passed.

**Review by dorin-iancu — DISMISSED** [review]

**Review by iulianpascalau — DISMISSED** [review]

File: vm/systemSmartContracts/esdt.go [hunk 7]
```diff
@@ -226,10 +230,10 @@ func (e *esdt) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 		return e.getContractConfig(args)
 	case "changeToMultiShardCreate":
 		return e.changeToMultiShardCreate(args)
-	case "setBurnForAll":
-		return e.setBurnForAll(args)
-	case "unsetBurnForAll":
-		return e.unSetBurnForAll(args)
+	case "setBurnRoleGlobally":
+		return e.setBurnRoleGlobally(args)
+	case "unsetBurnRoleGlobally":
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:235]: BUG: add `return e.unsetBurnRoleGlobally(args)`

File: vm/systemSmartContracts/esdt.go [hunk 8]
```diff
@@ -131,6 +133,7 @@ func NewESDTSmartContract(args ArgsNewESDTSmartContract) (*esdt, error) {
 		nftCreateONMultiShardEnableEpoch:  args.EpochConfig.EnableEpochs.ESDTNFTCreateOnMultiShardEnableEpoch,
 		metaESDTEnableEpoch:               args.EpochConfig.EnableEpochs.MetaESDTSetEnableEpoch,
 		registerAndSetAllRolesEnableEpoch: args.EpochConfig.EnableEpochs.ESDTRegisterAndSetAllRolesEnableEpoch,
+		checkMetaESDTOnRolesEnableEpoch:   args.EpochConfig.EnableEpochs.ManagedCryptoAPIsEnableEpoch,
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:136]: we need a new flag for this

File: vm/systemSmartContracts/esdt.go [hunk 9]
```diff
@@ -2177,8 +2181,11 @@ func (e *esdt) EpochConfirmed(epoch uint32, _ uint64) {
 	e.flagRegisterAndSetAllRoles.SetValue(epoch >= e.registerAndSetAllRolesEnableEpoch)
 	log.Debug("ESDT register and set all roles", "enabled", e.flagRegisterAndSetAllRoles.IsSet())
 
+	e.flagCheckMetaESDTOnRolesEnableEpoch.SetValue(epoch >= e.checkMetaESDTOnRolesEnableEpoch)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:2184]: where is this flag used?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/esdt.go (hunk 8 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:136]: this is not the change - line 135 is the change - github is showing it wrong

File: vm/systemSmartContracts/esdt.go (hunk 9 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:2184]: again - github is stupid. this is already in on master.

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**